File: AppLovin MAX Demo App - ObjC/AppLovin MAX Demo App - ObjC/MAX/MRECs/Table View/ALMAXMRecTableViewController.m
//
//  ALMAXMRecTableViewController.m
//  AppLovin MAX Demo App - ObjC
//
//  Created by Alan Cao on 6/30/22.
//  Copyright © 2022 AppLovin Corporation. All rights reserved.
//

#import "ALMAXMRecTableViewController.h"
#import "ALMAXMRecTableViewCell.h"
#import <AppLovinSDK/AppLovinSDK.h>

@interface ALMAXMRecTableViewController () <UITableViewDelegate, UITableViewDataSource, MAAdViewAdDelegate>

@property (nonatomic, weak) IBOutlet UITableView *tableView;

@property (nonatomic, strong) NSMutableArray<MAAdView *> *adViews;
@property (nonatomic, strong) NSMutableArray<NSString *> *sampleData;

@end

@implementation ALMAXMRecTableViewController
static const NSInteger kAdViewCount = 5;
static const NSInteger kAdInterval = 10;

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.adViews = [NSMutableArray arrayWithCapacity: kAdViewCount];
    self.sampleData = [NSMutableArray arrayWithArray: [UIFont familyNames]];
    
    // Configure table view
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    
    [self configureAdViews];
}

- (void)configureAdViews
{
    [self.tableView beginUpdates];
    
    // Insert rows at each interval to be used to display an ad
    for ( int i = 0; i < self.sampleData.count; i += kAdInterval )
    {
        [self.sampleData insertObject: @"" atIndex: i];
        [self.tableView insertRowsAtIndexPaths: @[[NSIndexPath indexPathForRow: i inSection: 0]] withRowAnimation: UITableViewRowAnimationAutomatic];
    }
    
    [self.tableView endUpdates];
    
    for ( int i = 0; i < kAdViewCount; i++ )
    {
        MAAdView *adView = [[MAAdView alloc] initWithAdUnitIdentifier: @"YOUR_AD_UNIT_ID" adFormat: MAAdFormat.mrec];
        adView.delegate = self;
        
        // Set this extra parameter to work around SDK bug that ignores calls to stopAutoRefresh()
        [adView setExtraParameterForKey: @"allow_pause_auto_refresh_immediately" value: @"true"];
        [adView stopAutoRefresh];
        
        // Load the ad
        [adView loadAd];
        [self.adViews addObject: adView];
    }
}

#pragma mark - UITableView

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.sampleData.count;
}

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath
{
    if ( indexPath.row % kAdInterval == 0 )
    {
        ALMAXMRecTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier: @"ALMAXMRecTableViewCell" forIndexPath: indexPath];
        
        // Select an ad view to display
        MAAdView *adView = self.adViews[(indexPath.row / kAdInterval) % kAdViewCount];
        
        // Configure cell with an ad
        [cell configureWithAdView: adView];
        
        return cell;
    }
    else
    {
        UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier: @"CustomCell" forIndexPath: indexPath];
        cell.textLabel.text = self.sampleData[indexPath.row];
        
        return cell;
    }
}

- (void)tableView:(UITableView *)tableView didEndDisplayingCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if ( [cell isKindOfClass: ALMAXMRecTableViewCell.class] )
    {
        [(ALMAXMRecTableViewCell *)cell stopAutoRefresh];
    }
}

#pragma mark - MAAdViewAdDelegate Protocol

- (void)didLoadAd:(MAAd *)ad {}

- (void)didFailToLoadAdForAdUnitIdentifier:(NSString *)adUnitIdentifier withError:(MAError *)error {}

- (void)didClickAd:(MAAd *)ad {}

- (void)didFailToDisplayAd:(MAAd *)ad withError:(MAError *)error {}

- (void)didExpandAd:(MAAd *)ad {}

- (void)didCollapseAd:(MAAd *)ad {}

#pragma mark - Deprecated Callbacks

- (void)didDisplayAd:(MAAd *)ad { /* DO NOT USE - THIS IS RESERVED FOR FULLSCREEN ADS ONLY AND WILL BE REMOVED IN A FUTURE SDK RELEASE */ }
- (void)didHideAd:(MAAd *)ad { /* DO NOT USE - THIS IS RESERVED FOR FULLSCREEN ADS ONLY AND WILL BE REMOVED IN A FUTURE SDK RELEASE */ }

@end

File: AppLovin MAX Demo App - Swift/AppLovin MAX Demo App - Swift/MAX/MRECs/ALMAXMRecTableViewController.swift
//
//  ALMAXMRecTableViewController.swift
//  AppLovin MAX Demo App - Swift
//
//  Created by Alan Cao on 6/27/22.
//  Copyright © 2022 AppLovin. All rights reserved.
//

import UIKit
import AppLovinSDK

class ALMAXMRecTableViewController: UIViewController
{
    private let kAdViewCount = 5
    private let kAdInterval = 10
    private var adViews: [MAAdView] = []
    private var sampleData: [String] = UIFont.familyNames
    
    @IBOutlet weak var tableView: UITableView!
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        
        // Configure table view
        tableView.delegate = self
        tableView.dataSource = self
        
        configureAdViews(count: kAdViewCount)
    }
    
    private func configureAdViews(count: Int)
    {
        tableView.beginUpdates()
        
        for i in stride(from: 0, to: sampleData.count, by: kAdInterval)
        {
            sampleData.insert("", at: i)
            tableView.insertRows(at: [IndexPath(row: i, section: 0)], with: .automatic)
        }
        
        tableView.endUpdates()
        
        for _ in 0 ..< count
        {
            let adView = MAAdView(adUnitIdentifier: "YOUR_AD_UNIT_ID", adFormat: .mrec)
            adView.delegate = self
            
            // Set this extra parameter to work around SDK bug that ignores calls to stopAutoRefresh()
            adView.setExtraParameterForKey("allow_pause_auto_refresh_immediately", value: "true")
            adView.stopAutoRefresh()
            
            // Load the ad
            adView.loadAd()
            adViews.append(adView)
        }
    }
}

// MARK: UITableView

extension ALMAXMRecTableViewController : UITableViewDelegate, UITableViewDataSource
{
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int
    {
        return sampleData.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell
    {
        
        if indexPath.row % kAdInterval == 0
        {
            let cell = tableView.dequeueReusableCell(withIdentifier: "ALMAXMRecTableViewCell", for: indexPath) as! ALMAXMRecTableViewCell
            
            // Select an ad view to display
            let adView = adViews[(indexPath.row / kAdInterval) % kAdViewCount]
            
            // Configure cell with an ad
            cell.configure(with: adView)
            
            return cell
        }
        else
        {
            let cell = tableView.dequeueReusableCell(withIdentifier: "customCell", for: indexPath)
            cell.textLabel!.text = sampleData[indexPath.row]
            
            return cell
        }
    }
    
    func tableView(_ tableView: UITableView, didEndDisplaying cell: UITableViewCell, forRowAt indexPath: IndexPath)
    {
        if let cell = cell as? ALMAXMRecTableViewCell
        {
            cell.stopAutoRefresh()
        }
    }
}

// MARK: MAAdViewAdDelegate Protocol

extension ALMAXMRecTableViewController: MAAdViewAdDelegate
{
    func didLoad(_ ad: MAAd) {}
    
    func didFailToLoadAd(forAdUnitIdentifier adUnitIdentifier: String, withError error: MAError) {}
    
    func didClick(_ ad: MAAd) {}
    
    func didFail(toDisplay ad: MAAd, withError error: MAError) {}
    
    func didExpand(_ ad: MAAd) {}
    
    func didCollapse(_ ad: MAAd) {}
    
    // MARK: Deprecated Callbacks
    
    func didDisplay(_ ad: MAAd) { /* DO NOT USE - THIS IS RESERVED FOR FULLSCREEN ADS ONLY AND WILL BE REMOVED IN A FUTURE SDK RELEASE */ }
    func didHide(_ ad: MAAd) { /* DO NOT USE - THIS IS RESERVED FOR FULLSCREEN ADS ONLY AND WILL BE REMOVED IN A FUTURE SDK RELEASE */ }
}
